### PR TITLE
Update TelemetryClient to enable tracking of dialog ('page') views

### PR DESCRIPTION
**Pull request by garypretty**
Addresses #3441 

Update TelemetryClient to enable tracking of page views and add automatic page view logging on dialog start for component / waterfall dialogs.

~~This PR now adds a new abstract class, ***LogTelemetryClient***, as well as an implementation of it within the ApplicationInsights library, ***AppInsightsTelemetryClient***. All current consumers of the existing ***BotTelemetryClient*** have been updated to use the new client, but leaving the existing functionality in place (marked as obselete) for backwards compatibility.~~

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClient.cs [hunk 1]
```diff
@@ -64,6 +64,14 @@ public interface IBotTelemetryClient
...
         void TrackTrace(string message, Severity severityLevel, IDictionary<string, string> properties);
 
+        /// <summary>
+        /// Logs a dialog entry / as an Application Insights page view.
+        /// </summary>
+        /// <param name="dialogName">The name of the dialog to log the entry / start for.</param>
+        /// <param name="properties">Named string values you can use to search and classify events.</param>
+        /// <param name="metrics">Measurements associated with this event.</param>
+        void TrackDialogView(string dialogName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null);
+
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder/IBotTelemetryClient.cs]: unfortunately, this is a breaking change.  You will need to add a IBotTelemetryClient2 or ITrackDialogView or something.

Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClient.cs]: we should switch to a BotTelemetryClient abstract base class

---
In reply to: [384199303](https://github.com/microsoft/botbuilder-dotnet/pull/3440#discussion_r384199303) [](ancestors = 384199303)

Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClient.cs]: good catch. Will re-visit this tomorrow.

File: libraries/Microsoft.Bot.Builder/LogTelemetryClient.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        {
+        }
+
+        /// <summary>
+        /// Logs a page view.
+        /// </summary>
+        /// <param name="name">The name of the page view to log.</param>
+        /// <param name="properties">Named string values you can use to search and classify events.</param>
+        /// <param name="metrics">Measurements associated with this event.</param>
+        public virtual void TrackPageView(string name, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null)
```
Comment (christopheranderson) [inline libraries/Microsoft.Bot.Builder/LogTelemetryClient.cs]: nit: naming - TrackPageView is leaking an AppInsights specific concept to the top level LogTelemetryClient, which doesn't seem right. It'd be great if this corresponded with a specific BF related user interaction.

File: libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs [hunk 3]
```diff
@@ -46,6 +59,15 @@ public TelemetryLoggerMiddleware(IBotTelemetryClient telemetryClient, bool logPe
         [JsonIgnore]
         public IBotTelemetryClient TelemetryClient { get; }
 
+        /// <summary>
+        /// Gets The telemetry client to send telemetry events to.
+        /// </summary>
+        /// <value>
+        /// The <see cref="IBotTelemetryClient"/> this middleware uses to log events.
```
Comment (christopheranderson) [inline libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs]: We're saying that `IBotTelemetryClient` is obsolete now, so this comment probably shouldn't refer to it?

File: libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs [hunk 4]
```diff
@@ -46,6 +59,15 @@ public TelemetryLoggerMiddleware(IBotTelemetryClient telemetryClient, bool logPe
         [JsonIgnore]
         public IBotTelemetryClient TelemetryClient { get; }
```
Comment (christopheranderson) [inline libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs]: Should this be obsolete now?

File: libraries/Microsoft.Bot.Builder/LogTelemetryClient.cs (hunk 2 above)
Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/LogTelemetryClient.cs]: Yeah, I went back a forth on this - I originally had it named TrackDialogView.  I think you are right - I'll probably go back.

File: libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs (hunk 3 above)
Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs]: I was marking it as obselete to indicate that a user that they should update their config but leaving in place so that existing solutions still work - based on our discussion, I agree that we should remove the obselete flag though until R9. Give us change to update docs / samples / advice over that period and get our updated telemetry story out there.

File: libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs (hunk 4 above)
Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/TelemetryLoggerMiddleware.cs]: marking it as obselete to indicate that a user that they should update their config but leaving in place so that existing solutions still work

**Comment by garypretty** [comment]
Agreed with @christopheranderson that, providing we go ahead with this approach, we will not obselete the existing telemetry bits until R9 - this will allow us to run the two side by side until all of our messaging etc is up to date and also to provide us with a chance to ensure that the new bits meet the demands of adaptive too.

This also presents the possibility of us deferring the JS and other repo updates to R9 as well if we don't get through them all next week.

**Comment by garypretty** [comment]
**Update: ** Due to the fact that if we pursue a non-breaking change then 100% of people who have telemetry enabled within their existing bots will need to make changes to migrate to an updated configuration, coupled with the fact that it is very likely that the number of people to extend the IBotTelemetryClient interface within their own solution is incredibly small - the decision has been taken to extend the interface. 
This allows us to allow the vast majority of existing telemetry users to gain the new capabilities without making any changes within their solution.  It also does not preclude us from replacing the interface in the future.

cc @cleemullins / @scheyal

**Comment by garypretty** [comment]
Implemented additional interface / extension method to avoid breaking change as discussed with @johnataylor

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    public static class IBotTelemetryClientExtensions
+    {
+        /// <summary>
+        /// Adds the ability to call TrackPageView on the TelemetryClient if it implements IBotPageViewTelemetryClient as well as IBotTelemetryClient.
+        /// </summary>
+        /// <param name="telemetryClient">The TelemetryClient that implements IBotTelemetryClient.</param>
+        /// <param name="dialogName">The name of the dialog to log the entry / start for.</param>
+        /// <param name="properties">Named string values you can use to search and classify events.</param>
+        /// <param name="metrics">Measurements associated with this event.</param>
+        public static void TrackPageView(this IBotTelemetryClient telemetryClient, string dialogName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null)
```
Comment (johnataylor) [inline libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: why not call this extension method TrackDialog? isn't that what you original intended?

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            if (telemetryClient is IBotPageViewTelemetryClient pageViewClient)
+            {
+                pageViewClient.TrackPageView(dialogName, properties, metrics);
+            }
+            else
+            {
+                telemetryClient.TrackTrace("TelemetryClient cannot track PageView telemtry as it does not implement IBotPageViewTelemetryClient", Severity.Information, properties);
```
Comment (johnataylor) [inline libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: can't we gently "fall back" to less sophisticated behavior (fro example, an event) rather than abruptly just trace saying "update your stuff."

I'm thinking the user of the telemetry functionality is not necessary the implementer of the telemetry provider. If the user is constrained for some reason to some particular telemetry provider (other than our appsignts solution) then they at least get some reasonable behavior. Or am I miss-representing the scenario?

File: tests/Microsoft.Bot.Builder.Dialogs.Tests/WaterfallTests.cs [hunk 7]
```diff
@@ -466,6 +466,11 @@ public void TrackDependency(string dependencyTypeName, string target, string dep
                 throw new NotImplementedException();
             }
 
+            public void TrackPageView(string dialogName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null)
```
Comment (johnataylor) [inline tests/Microsoft.Bot.Builder.Dialogs.Tests/WaterfallTests.cs]: and this isn't a realistic scenario right, just testing? we are suggesting if you _don't_ implement TrackPageView then you _don't_ implement IBotPageViewTelemetryClient

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs (hunk 6 above)
Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: Surely if the user is using a different system then they will implement their own class that uses both interfaces?  In the case of our method we want to handle the situation where the page view is not logged because the interface has not been implemented.  I think a warning here is fine and it is probably an edge case.

The alternative would be to log additional fallback events using TrackEvent, but the danger then is that you start polluting someone's telemetry and this will only be in the case they have their own implementation. So, it feels better to warn them gracefully rather than explicitly log something.

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs (hunk 5 above)
Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: I though we had agreed to fall back to the more native event names. Given this is an extension to and therefore used like it is on IBotTelemetryClient, then we should name it however we would name it if we were just extending the interface.

File: tests/Microsoft.Bot.Builder.Dialogs.Tests/WaterfallTests.cs (hunk 7 above)
Comment (garypretty) [reply tests/Microsoft.Bot.Builder.Dialogs.Tests/WaterfallTests.cs]: Agreed. I will remove.

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs (hunk 5 above)
Comment (johnataylor) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: I was seeing this as layered. Lower layer being derived directly from the raw capabilities of the service we are abstracting. This lower layer is captured in a series of interfaces. (Used to be one interface, now it's two.)

The higher layer is more about the usage pattern in the application. For example, TrackDialog corresponds to the concept of Dialogs that exist only with the application and not the telemetry infrastructure. The implementation of the provider happens to map this to a PageView because we felt that sort of fitted OK.

The benefit of the multiple interfaces was that this application concept of Dialog could actually be mapped differently in a less capable (aka legacy) telemetry provider. This negotiation (implemented with a runtime type check) facilitates the back compat and graceful fallback.

Anyhow the abstractions TrackDialog and TrackPageView target different audiences. The TRackPageView being intended for someone implementing something like TrackDialog.

Anyhow that was my rational. I didn't see the extension method as extending the interface, more as a higher level abstraction, as if we had always intended it to work this way.

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs (hunk 6 above)
Comment (johnataylor) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: Given we shipped an interface, we implied there can be other implementations. So assuming there are, then those preexisting implementations don't have a TrackPageView concept the TrackEvent is the best thing that can happen.

I'm assuming a world where someone has preexisting telemetry and I just wanted to get the data into their telemetry rather than a warning message telling them that they must upgrade their telemetry capabilities.

In other words, the user of the telemetry client is in control, if they want the enhanced experience of pageview then they can upgrade their telemetry provider to get that. In the meantime they get an event.

(Having said that I don't have any personal experience of what it means to get a PageView rather than an Event in my telemetry. I'm assuming its better, but I'm also assuming an Event kinda works.)

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs (hunk 5 above)
Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: Ok - I am fine either way and do see your side of the argument too. I will alter the naming. Ironically, that is how I had named it when this PR first got opened :)

Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: Updated extension method to DialogView as discussed.

File: libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs (hunk 6 above)
Comment (garypretty) [reply libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs]: Updated Trace event to contain info as opposed to warning message as discussed.
